File: brain.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;


class input_size{
    public:
    int a,b,e,f,g;
    void inputone(){
    cout << "Input image size: \n (We recommend 640 by 590!)" << endl;
    cin >> a >> b;
    cout << "Updated Image size" << endl;
    }

    void inputtwo(){
        cout << "Input Pixel colors for contrast: \n (We recommend 255,255,255!)" << endl;
        cin >> e >> f >> g;
        cout << "Updated pixel colors" << endl;
    }
};

int main() {

    input_size obj1, obj2;
    obj1.inputone();
    obj2.inputtwo();
    int c=obj1.a, d=obj1.b, p=obj2.e, q=obj2.f, r=obj2.g;
    // Load input image
    Mat image = imread("C:/Users/Rishi/OneDrive/Pictures/brain.png");

    // Resize image to a smaller size
    Size newSize(640, 590);
    resize(image, image, newSize);

    // Convert image to grayscale
    Mat grayImage;
    cvtColor(image, grayImage, COLOR_BGR2GRAY);

    // Apply top hat filter
    Mat kernel = getStructuringElement(MORPH_ELLIPSE, Size(15, 15));
    Mat topHat;
    morphologyEx(grayImage, topHat, MORPH_TOPHAT, kernel);

    // Blur image
    GaussianBlur(topHat, topHat, Size(5, 5), 0);

    // Apply thresholding for segmenting
    Mat thresholded;
    threshold(topHat, thresholded, 0, 255, THRESH_BINARY + THRESH_OTSU);

    // Compute distance transform
    thresholded.convertTo(thresholded, CV_8UC1);

    Mat dist;
    distanceTransform(thresholded, dist, DIST_L2, 3);

    // Find local maxima in distance transform
    Mat localMaxima;
    dilate(thresholded, localMaxima, Mat());
    compare(thresholded, localMaxima, localMaxima, CMP_EQ);
    localMaxima.convertTo(localMaxima, CV_8UC1);

    // Find markers for watershedding
    Mat markers;
    connectedComponents(localMaxima, markers);

    // Apply watershedding
    watershed(image, markers);

    // Display output
    Mat output = Mat::zeros(image.size(), CV_8UC3);
    for (int i = 0; i < markers.rows; i++) {
        for (int j = 0; j < markers.cols; j++) {
            int index = markers.at<int>(i, j);
            if (index == -1) {
                output.at<Vec3b>(i, j) = Vec3b(255, 255, 255);
            } else if (index == 1) {
                output.at<Vec3b>(i, j) = Vec3b(p, q, r);
            }
        }
    }
    imshow("Original Image", image);
    imshow("Thresholded Image", thresholded);
    imshow("Output", output);
    waitKey(0);

    string originalFileName = "original.png";
    imwrite(originalFileName, image);

    string thresholdedFileName = "thresholded.png";
    imwrite(thresholdedFileName, thresholded);

    string outputFileName = "output.png";
    imwrite(outputFileName, output);
    return 0;
}

File: brain.py
# Recreating in Python

import cv2
import numpy as np

class InputSize:
    def input_one(self):
        print("Input image size: (recommended 640 by 590)")
        self.a, self.b = map(int, input().split())
        print("Updated Image size")

    def input_two(self):
        print("Input Pixel colors for contrast: (recommended 255,255,255)")
        self.e, self.f, self.g = map(int, input().split())
        print("Updated pixel colors")

def main():
    obj1 = InputSize()
    obj2 = InputSize()
    obj1.input_one()
    obj2.input_two()
    c, d, p, q, r = obj1.a, obj1.b, obj2.e, obj2.f, obj2.g
    
    # Load input image
    image = cv2.imread("C:/Users/Rishi/OneDrive/Pictures/brain.png")
    
    # Resize image
    newSize = (c, d)
    image = cv2.resize(image, newSize)
    
    # Convert image to grayscale
    grayImage = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    
    # Apply top hat filter
    kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (15, 15))
    topHat = cv2.morphologyEx(grayImage, cv2.MORPH_TOPHAT, kernel)
    
    # Blur image
    topHat = cv2.GaussianBlur(topHat, (5, 5), 0)
    
    # Apply thresholding for segmenting
    _, thresholded = cv2.threshold(topHat, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
    
    # Compute distance transform
    thresholded = np.uint8(thresholded)
    dist = cv2.distanceTransform(thresholded, cv2.DIST_L2, 3)
    
    # Find local maxima in distance transform
    localMaxima = cv2.dilate(thresholded, None)
    localMaxima = (thresholded == localMaxima)
    localMaxima = np.uint8(localMaxima)
    
    # Find markers for watershedding
    _, markers = cv2.connectedComponents(localMaxima)
    
    # Apply watershedding
    markers = cv2.watershed(image, markers)
    
    # Display output
    output = np.zeros(image.shape, dtype=np.uint8)
    for i in range(markers.shape[0]):
        for j in range(markers.shape[1]):
            index = markers[i, j]
            if index == -1:
                output[i, j] = [255, 255, 255]
            elif index == 1:
                output[i, j] = [p, q, r]
    
    cv2.imshow("Original Image", image)
    cv2.imshow("Thresholded Image", thresholded)
    cv2.imshow("Output", output)
    cv2.waitKey(0)
    
    originalFileName = "./output/original.png"
    cv2.imwrite(originalFileName, image)
    
    thresholdedFileName = "./output/thresholded.png"
    cv2.imwrite(thresholdedFileName, thresholded)
    
    outputFileName = "./output/output.png"
    cv2.imwrite(outputFileName, output)

if __name__ == "__main__":
    main()
